File: src/main/java/gg/mineral/practice/category/Category.java
package gg.mineral.practice.category;

import org.bukkit.inventory.ItemStack;

import gg.mineral.api.collection.GlueList;
import gg.mineral.api.config.FileConfiguration;
import gg.mineral.practice.gametype.Gametype;
import gg.mineral.practice.managers.CategoryManager;
import gg.mineral.practice.managers.QueuetypeManager;
import gg.mineral.practice.queue.QueuetypeMenuEntry;
import gg.mineral.practice.util.SaveableData;
import gg.mineral.practice.util.items.ItemStacks;
import lombok.Getter;
import lombok.val;

public class Category implements SaveableData, QueuetypeMenuEntry {

    final FileConfiguration config = CategoryManager.getConfig();

    @Getter
    ItemStack displayItem;
    @Getter
    String displayName;
    @Getter
    private final String name;
    @Getter
    GlueList<Gametype> gametypes = new GlueList<>();
    final String path;
    private static final String ENABLED_STRING = ".Enabled";

    public Category(String name) {
        this.name = name;
        this.path = "Category." + getName() + ".";
    }

    @Override
    public boolean isBotsEnabled() {
        for (val gametype : gametypes)
            if (gametype.isBotsEnabled())
                return true;
        return false;
    }

    public void setDisplayItem(ItemStack displayItem) {
        this.displayItem = displayItem;
        save();
    }

    public void setDisplayName(String displayName) {
        this.displayName = displayName;
        save();
    }

    public void addGametype(Gametype gametype) {
        gametypes.add(gametype);
        save();
    }

    public void removeGametype(Gametype gametype) {
        gametypes.remove(gametype);
        save();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Category category)
            return category.getName().equalsIgnoreCase(getName());
        return false;
    }

    @Override
    public int hashCode() {
        return getName().hashCode();
    }

    @Override
    public void save() {
        for (val q : QueuetypeManager.getQueuetypes().values()) {

            if (!q.getMenuEntries().containsKey(this)) {
                config.set(path + q.getName() + ENABLED_STRING, false);
                continue;
            }

            int slot = q.getMenuEntries().getInt(this);

            config.set(path + q.getName() + ENABLED_STRING, true);
            config.set(path + q.getName() + ".Slot", slot);
        }

        config.set(path + "DisplayName", displayName);
        config.set(path + "DisplayItem", displayItem);

        config.save();
    }

    @Override
    public void load() {
        this.displayItem = config.getItemstack(path + "DisplayItem",
                ItemStacks.DEFAULT_CATEGORY_DISPLAY_ITEM);
        this.displayName = config.getString(path + "DisplayName", getName());

        for (val q : QueuetypeManager.getQueuetypes().values())
            if (config.getBoolean(path + q.getName() + ENABLED_STRING, false))
                q.getMenuEntries().put(this, (int) config.getInt(path + q.getName() + ".Slot", 0));

    }

    @Override
    public void setDefaults() {
        this.displayItem = ItemStacks.DEFAULT_CATEGORY_DISPLAY_ITEM;
        this.displayName = getName();

        for (val q : QueuetypeManager.getQueuetypes().values())
            if (config.getBoolean(path + q.getName() + ENABLED_STRING, false))
                q.getMenuEntries().put(this, 0);
    }

    @Override
    public void delete() {
        config.remove("Category." + getName());
        config.save();
    }
}

File: src/main/kotlin/gg/mineral/practice/category/Category.kt
package gg.mineral.practice.category

import gg.mineral.practice.gametype.Gametype
import gg.mineral.practice.managers.CategoryManager
import gg.mineral.practice.managers.GametypeManager
import gg.mineral.practice.queue.QueuetypeMenuEntry
import gg.mineral.practice.util.config.ItemStackProp
import gg.mineral.practice.util.config.StringProp
import gg.mineral.practice.util.config.yaml.FileConfiguration
import gg.mineral.practice.util.items.ItemStacks
import it.unimi.dsi.fastutil.bytes.ByteOpenHashSet

class Category(val name: String, val id: Byte) : QueuetypeMenuEntry {
    val config: FileConfiguration = CategoryManager.config
    val path: String = "Category.$name."
    override var displayItem by ItemStackProp(config, path + "DisplayItem", ItemStacks.DEFAULT_CATEGORY_DISPLAY_ITEM)
    override var displayName by StringProp(config, path + "DisplayName", name)
    override val botsEnabled: Boolean
        get() {
            val iterator = gametypes.iterator()
            while (iterator.hasNext()) {
                val gametypeId = iterator.nextByte()
                if (GametypeManager.gametypes[gametypeId].botsEnabled) return true
            }
            return false
        }

    val gametypes by lazy {
        val set = ByteOpenHashSet()
        for (gametype in GametypeManager.gametypes.values)
            if (gametype.inCategory && gametype.categoryName.equals(name, ignoreCase = true)) set.add(gametype.id)
        set
    }

    fun addGametype(gametype: Gametype) {
        gametypes.add(gametype.id)

        // TODO: change how this operates
        gametype.inCategory = true
        gametype.categoryName = name
    }

    fun removeGametype(gametype: Gametype) {
        gametypes.remove(gametype.id)
        // TODO: change how this operates
        gametype.inCategory = false
    }

    override fun equals(other: Any?): Boolean {
        if (other is Category) return other.name.equals(this.name, ignoreCase = true)
        return false
    }

    override fun hashCode() = name.hashCode()

    fun delete() {
        config.remove("Category.$name")
        config.save()
    }
}
